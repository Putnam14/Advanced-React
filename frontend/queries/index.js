import gql from 'graphql-tag';
import { itemDetails } from './fragments';

export const CREATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation createItem($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
      ...itemDetails
    }
  }
`;

export const SIGNUP_MUTATION = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const SIGNIN_MUTATION = gql`
  mutation signup($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const REQUEST_RESET_MUTATION = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      id
    }
  }
`;

export const RESET_MUTATION = gql`
  mutation resetPassword($resetToken: String!, $password: String!, $confirmPassword: String!) {
    resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const CREATE_ORDER_MUTATION = gql`
  mutation CreateOrderMutation($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

export const ALL_ITEMS_QUERY = gql`
  # Import the Fragment
  ${itemDetails}
  query AllItemsQuery($skip: Int = 0, $first: Int = 2) {
    itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "itemsConnection") {
      aggregate {
        count
      }
    }
    items(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "items") {
      ...itemDetails
    }
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  ${itemDetails}
  query FindItem($id: ID!) {
    items(where: { id: $id }) {
      user {
        id
        email
        name
      }
      ...itemDetails
    }
  }
`;

export const SINGLE_ORDER_QUERY = gql`
  query order($id: ID!) {
    order(where: { id: $id }) {
      id
      charge
      total
      createdAt
      items {
        id
        title
        price
        description
        image
        quantity
      }
    }
  }
`;

export const SEARCH_ITEMS_QUERY = gql`
  ${itemDetails}
  query SearchItems($searchTerm: String!) {
    items(
      where: {
        OR: [
          { title_contains: $searchTerm }
          { description_contains: $searchTerm }
        ]
      }
    ) {
      ...itemDetails
    }
  }
`;

export const REMOVE_ITEM_MUTATION = gql`
  mutation RemoveItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
      description
    }
  }
`;

export const UPDATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation UpdateItem(
    $id: ID!
    $title: String
    $description: String
    $price: Int
  ) {
    updateItem(
      id: $id
      description: $description
      title: $title
      price: $price
    ) {
      ...itemDetails
    }
  }
`;

export const UPDATE_USER_MUTATION = gql`
  mutation UpdateUser($name: String!) {
    updateUser(name: $name) {
      name
    }
  }
`;

export const CURRENT_USER_QUERY = gql`
  ${itemDetails}
  query userQuery {
    me {
      id
      email
      name
      orders {
        charge
        id
        total
      }
      cart {
        id
        quantity
        item {
          ...itemDetails
        }
      }
    }
  }
`;

export const USER_ORDERS_QUERY = gql`
  query orders {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      updatedAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      item {
        id
        price
        description
        image
        title
      }
    }
  }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

export const ALL_USERS_QUERY = gql`
  query Users {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;
