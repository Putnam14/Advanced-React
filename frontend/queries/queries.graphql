fragment itemDetails on Item {
  __typename
  id
  title
  price
  description
  image
  largeImage
}

mutation CREATE_ITEM_MUTATION($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
  createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
    ...itemDetails
  }
}

mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
  signup(email: $email, name: $name, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation SIGNIN_MUTATION($email: String!, $password: String!) {
  signin(email: $email, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation REQUEST_RESET_MUTATION($email: String!) {
  requestReset(email: $email) {
    id
  }
}

mutation RESET_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
  resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation CREATE_ORDER_MUTATION($token: String!) {
  createOrder(token: $token) {
    id
    charge
    total
    items {
      id
      title
    }
  }
}

query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
  itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    aggregate {
      count
    }
  }
  items(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    ...itemDetails
  }
}

query SINGLE_ITEM_QUERY($id: ID!) {
  items(where: { id: $id }) {
    user {
      id
      email
      name
    }
    ...itemDetails
  }
}

query SINGLE_ORDER_QUERY($id: ID!) {
  order(id: $id) {
    id
    charge
    total
    createdAt
    user {
      id
    }
    items {
      id
      title
      price
      description
      image
      quantity
    }
  }
}

query SEARCH_ITEMS_QUERY($searchTerm: String!) {
  items(
    where: {
      OR: [
        { title_contains: $searchTerm }
        { description_contains: $searchTerm }
      ]
    }
  ) {
    ...itemDetails
  }
}

mutation REMOVE_ITEM_MUTATION($id: ID!) {
  deleteItem(id: $id) {
    id
    title
    description
  }
}

mutation UPDATE_ITEM_MUTATION(
  $id: ID!
  $title: String
  $description: String
  $price: Int
) {
  updateItem(
    id: $id
    description: $description
    title: $title
    price: $price
  ) {
    ...itemDetails
  }
}

mutation UPDATE_USER_MUTATION($name: String!) {
  updateUser(name: $name) {
    name
  }
}

query CURRENT_USER_QUERY {
  me {
    id
    email
    name
    permissions
    orders {
      id
      charge
      total
    }
    cart {
      id
      quantity
      item {
        ...itemDetails
      }
    }
  }
}

query USER_ORDERS_QUERY {
  orders(orderBy: createdAt_DESC) {
    id
    total
    createdAt
    updatedAt
    items {
      id
      title
      price
      description
      quantity
      image
    }
  }
}

mutation ADD_TO_CART_MUTATION($id: ID!) {
  addToCart(id: $id) {
    id
    quantity
    item {
      id
      price
      description
      image
      title
    }
  }
}

mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
  removeFromCart(id: $id) {
    id
  }
}

query ALL_USERS_QUERY {
  users {
    id
    name
    email
    permissions
  }
}

mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
  updatePermissions(permissions: $permissions, userId: $userId) {
    id
    permissions
    name
    email
  }
}

query LOCAL_STATE_QUERY {
  cartOpen @client
}

mutation TOGGLE_CART_MUTATION {
  toggleCart @client
}

