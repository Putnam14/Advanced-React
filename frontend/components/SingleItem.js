import { Query } from 'react-apollo';
import PropTypes from 'prop-types';
import { SINGLE_ITEM_QUERY } from '../queries/queries';
import styled from 'styled-components';
import Link from 'next/link';
import Dump from './Dump';
import Error from './ErrorMessage';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  img {
    width: 100%;
    object-fit: cover;
  }
  .details {
    margin: 3rem;
    font-size: 3rem;
  }
`;

const SingleItem = props => (
  <Query query={SINGLE_ITEM_QUERY} variables={{ id: props.id }}>
    {({ data, loading, error }) => {
      if (loading) return <p>Loading...</p>;
      if (error) return <Error error={error} />;
      const [item] = data.items;
      return (
        <SingleItemStyles data-test="SingleItem">
          <img src={item.largeImage || item.image} alt={item.title} />
          <div className="details">
            <h2>Viewing {item.title}</h2>
            <p>{item.description}</p>
            <Link
              href={{
                pathname: '/update',
                query: { id: item.id },
              }}
            >
              <a>Edit ✏️</a>
            </Link>
          </div>
        </SingleItemStyles>
      );
    }}
  </Query>
);

SingleItem.propTypes = {
  id: PropTypes.string.isRequired,
};

export default SingleItem;
