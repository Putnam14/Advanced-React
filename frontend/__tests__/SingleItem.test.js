import React from 'react';
import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import SingleItem from '../components/SingleItem';
import { SINGLE_ITEM_QUERY } from '../queries/queries';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from './mockMang';

const data = {
  items: [fakeItem()],
};

describe('<SingleItem/>', () => {
  it('Renders with Proper Data', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        delay: 50,
        result: { data },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    // Check for loading state
    expect(wrapper.text()).toContain('Loading...');

    // wait for response and refresh
    await wait(55);
    wrapper.update();
    // Grab the piece we want
    const Item = wrapper.find('[data-test="SingleItem"]');
    // snapshot it!
    expect(toJSON(Item)).toMatchSnapshot();
  });

  it('Errors with a not found Item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        delay: 50,
        // we can return data and an error
        result: { data, errors: [{ message: 'Item not found!' }] },
        // or just an error!
        // error: new Error('offline!'),
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );

    // wait for response
    await wait(55);
    wrapper.update();

    const Item = wrapper.find('[data-test="graphql-error"]');
    console.log(Item.debug());
    expect(toJSON(Item)).toMatchSnapshot();
  });
});
