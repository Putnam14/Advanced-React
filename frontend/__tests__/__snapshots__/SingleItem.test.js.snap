// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> Errors with a not found Item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found!
</p>
`;

exports[`<SingleItem/> Renders with Proper Data 1`] = `
Array [
  <styled.div
    data-test="SingleItem"
  >
    <div
      className="sc-bwzfXH dFvSZo"
      data-test="SingleItem"
    >
      <img
        alt="dogs are best"
        src="dog.jpg"
      />
      <div
        className="details"
      >
        <h2>
          Viewing 
          dogs are best
        </h2>
        <p>
          dogs
        </p>
        <Link
          href={
            Object {
              "pathname": "/update",
              "query": Object {
                "id": "123",
              },
            }
          }
        >
          <a
            href="/update?id=123"
            onClick={[Function]}
          >
            Edit ✏️
          </a>
        </Link>
      </div>
    </div>
  </styled.div>,
  <div
    className="sc-bwzfXH dFvSZo"
    data-test="SingleItem"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      className="details"
    >
      <h2>
        Viewing 
        dogs are best
      </h2>
      <p>
        dogs
      </p>
      <Link
        href={
          Object {
            "pathname": "/update",
            "query": Object {
              "id": "123",
            },
          }
        }
      >
        <a
          href="/update?id=123"
          onClick={[Function]}
        >
          Edit ✏️
        </a>
      </Link>
    </div>
  </div>,
]
`;
